### DISCUSSION TITLE
useQuery used only key => "Missing queryFn" error

### DISCUSSION THREAD

**Discussion by gronxb** [TanStack/query#3407] [Q&A] [answered]
Hello, I need your help:sob:

I got the data by setting queryFn in queryClient with setQueryDefaults and setting only the key in useQuery from the child components. In this case, it will be good work, but the "Missing queryFn error" is being output to the console.   Can't I just use the only key for useQuery?

case:
```js

//... querys
queryClient.setQueryDefaults(["user"], {
  queryFn: () => fetchPromise("user")
});

// ... components used
const {data} = useQuery(["user"]); // this work ! but console.error
```

**Comment by TkDodo** [comment, marked as answer]
this should work. please provide a codesandbox reproduction showing the issue, thanks.

**Reply by gronxb** [reply]
Thank you!! I mistake during implementation websocket query.

**Reply by samiUllah1526** [reply]
I am facing the same issue but I don't know what I'm missing. can you please looking into it?

When use only axios its return the data but when I use react query it says **"Missing queryFn"**

### API function
`const getTasks = async () => {
  const response = await privateApi.get("/tasks", getAuthorizationHeader());
  return response.data;
};
`

### Query
`
  const {
    isLoading,
    isError,
    error,
    data: tasksdata,
  } = useQuery("tasks", getTasks);
  
<img width="626" alt="Screen Shot 2022-07-23 at 6 32 10 PM" src="https://user-images.githubusercontent.com/85010922/180607250-1a9b6df9-15a6-4238-88d5-9e28bbf99571.png">

  console.log({ isLoading, isError, error, tasksdata });
`

**Reply by vezaynk** [reply]
@TkDodo Would it be possible to wrap the returned errors of this kind in an `Error` object?

**Reply by TkDodo** [reply]
Yes, we are doing that in v5

**Comment by jhoanborges** [comment]
+1 same issue

**Reply by frederikhors** [reply]
Are you using v4 or v5?

**Comment by hamedtahmasebi** [comment]
Experiencing the same issue in
`"@tanstack/react-query": "^5.56.2",`

I am using the query in a container ( page level ) with the queryFn, and using it only with keys in subcomponents,


The main page:
```
const Page = () => {

  const { refetch, error } = useQuery({
        queryFn: async () => {
            return await HistoryApi.getHistoryData(filterOptions)
        },
        queryKey: [HistoryApi.QueryKeys.GET_HISTORY_DATA],
        retry: failureCount => failureCount < 3,
    })
    return (<>...</>)
}
```


subcomponents:
```
const SubComp = () => {

    const { data: historyData, isFetching: isFetchingHistoryData } = useQuery({
        // queryFn: async () => {
        //     return await HistoryApi.getHistoryData(filterOptions)
        // },
        queryKey: [HistoryApi.QueryKeys.GET_HISTORY_DATA],
    })
    return <>...</>
}
```

```
Error: Missing queryFn: '["GET_HISTORY_DATA"]'
    at utils.ts:418:22
    at Object.fetchFn [as fn] (query.ts:422:14)
    at run (retryer.ts:157:49)
    at Object.start (retryer.ts:222:9)
    at _a4.fetch (query.ts:539:26)
    at _a9.executeFetch_fn (queryObserver.ts:337:71)
    at _a9.onSubscribe (queryObserver.ts:98:9)
    at _a9.subscribe (subscribable.ts:11:10)
    at useBaseQuery.ts:85:22
    at subscribeToStore (react-dom.development.js:16139:10)
```

**Reply by TkDodo** [reply]
Don't use it without a queryFn. Where do you have this pattern from ?

**Reply by vendramini** [reply]
If I want to use react query to store and share local data that has no queryFn, what would be the correct pattern?

I'm doing like this:
```
const query = useQuery({
    queryKey: ['.environment'],
    queryFn: () => EXPO_PUBLIC_ENV,
    gcTime: Number.MAX_VALUE,
    staleTime: Number.MAX_VALUE,
    meta: { persist: true },
  });
```
I was using `placeholderData` but it complains about `queryFn` missing, so I changed like that. Apparently it's working, but I don't know if it is the best way to do it.

**Reply by TkDodo** [reply]
> If I want to use react query to store and share local data

This is not recommended, so there is no “best way” to do it. The recommendation is to use a client-state manager like zustand, redux, jotai, valtio, useState + react context, whatever.. but not react-query.

**Reply by vendramini** [reply]
Are there technical reasons to not be recommended? Because honestly I've been doing it for a long time and it's been working pretty well, I've never had problems. It makes my work much easier than having to deal with 2 different approaches - also, demanding less third-party libraries which means less code maintenance. Specially with react-native, using the plugin to store data locally is really handy.

**Reply by TkDodo** [reply]
query is an async state manager, not a synchronous state manager. We do batching and revalidation and have timers for async needs. Not only do you need to come up with a unique key for everything and have your environment query data be of type `string | undefined` instead of just `string`, there might also be one render cycle where it’s undefined at runtime, followed by a render cycle where data from the queryFn is returned. Are there ways around that? Sure, but your code isn’t doing it. 

Same as sticking data from an API request into manual redux with thunks was a bad idea, so is doing the opposite - putting everything into react-query. Synchronous state and Asynchronous state have different needs, that’s why we recommend different state managers for them.

Mandatory watch from 4 years ago, still gold: https://www.youtube.com/watch?v=seU46c6Jz7E

**Reply by vendramini** [reply]
Sorry delay, I've been too busy.

You are right, that video explains a lot.

So, the ideally flow would be to fetch X from server, copy its content to some local store management (Zustand, Context, jotai, useState), do whatever we need to do, update the server then update react-query again if needed?

**Reply by guseggert** [reply]
This pattern is also recommended by jotai-tanstack-query, see ex https://github.com/jotaijs/jotai-tanstack-query?tab=readme-ov-file#incremental-adoption

**Comment by Meligy** [comment]
It might be worth noting that one of the official examples in the docs is throwing this error:
React Example: Nextjs App Prefetching
https://tanstack.com/query/latest/docs/framework/react/examples/nextjs-app-prefetching
(I'm unable to run it on the page itself, but clicking the Stackblitz link brings it)

The example seems fairly innocent. I'm unable to tell what might be wrong.

The stack trace seems to mention the hydration boundary, but nothing more than that.

**Reply by TkDodo** [reply]
The example doesn't work in stackblitz. If you check it out and run it locally, it works fine.
